oci/oci.go

```diff
@@ -0,0 +1,110 @@
+package oci
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	mutils "github.com/projectatomic/skopeo/manifest"
+	"github.com/projectatomic/skopeo/types"
+)
+
+type descriptor struct {
+	Digest    string `json:"digest"`
+	MediaType string `json:"mediaType"`
+	Size      int64  `json:"size"`
+}
+
+type ociImageDestination struct {
+	dir string
+}
+
+// NewOCIImageDestination returns an ImageDestination for writing to an existing directory.
+func NewOCIImageDestination(dir string) types.ImageDestination {
+	return &ociImageDestination{dir: dir}
+}
+
+func (d *ociImageDestination) CanonicalDockerReference() (string, error) {
+	return "", fmt.Errorf("Can not determine canonical Docker reference for an OCI directory")
+}
+
+func (d *ociImageDestination) PutManifest(manifest []byte) error {
+	refsPath := filepath.Join(d.dir, "refs")
+	if _, err := os.Stat(refsPath); err != nil && os.IsNotExist(err) {
+		if err := os.MkdirAll(refsPath, 0777); err != nil {
+			return err
+		}
+	}
+	digest, err := mutils.Digest(manifest)
+	if err != nil {
+		return err
+	}
+	desc := descriptor{}
+	desc.Digest = digest
+	desc.MediaType = "application/vnd.oci.image.manifest.v1+json"
+	desc.Size = int64(len(manifest))
+	data, err := json.Marshal(desc)
+	if err != nil {
+		return err
+	}
+	// TODO(runcom):
+	// make a function to create an OCI manifest from a distribution one
+	// convert the manifest to OCI v1 and first write it to ./blobs/sha256-$(manifest digest)
+
+	if err := ioutil.WriteFile(blobPath(d.dir, digest), manifest, 0644); err != nil {
+		return err
+	}
+	// TODO(runcom): ugly here?
+	if err := ioutil.WriteFile(ociLayoutPath(d.dir), []byte(`{"imageLayoutVersion": "1.0.0"}`), 0644); err != nil {
+		return err
+	}
+	// TODO(runcom): we should have a reference name when writing manifest here...
+	return ioutil.WriteFile(manifestPath(d.dir, "TODO"), data, 0644)
+}
+
+func (d *ociImageDestination) PutBlob(digest string, stream io.Reader) error {
+	blobsPath := filepath.Join(d.dir, "blobs")
+	if _, err := os.Stat(blobsPath); err != nil && os.IsNotExist(err) {
+		if err := os.MkdirAll(blobsPath, 0777); err != nil {
+			return err
+		}
+	} else if err != nil {
+		return err
+	}
```

Comment:
It seems this could be clearer with an `ensureParentDirectoryExists` helper function, instead of duplicating the computation of `blobsPath` here and `blobPath` below, and duplicating the `Stat`/`IsNotExist`/`MkdirAll` logic.